### PR TITLE
Restructure amount validation

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: src/main/app/forms/validation/validators/fractions.ts [hunk 1]
```diff
@@ -29,6 +29,10 @@ export class FractionsConstraint implements ValidatorConstraintInterface {
       return true
     }
 
+    if (value < 0) {
```
Comment (robertparkinson) [inline src/main/app/forms/validation/validators/fractions.ts]: refactor this validator. Too many return trues. Combine in one method or use a switch statement.

File: src/main/app/forms/validation/validators/isLowerThan.ts [hunk 2]
```diff
@@ -25,6 +25,10 @@ export class IsLowerThanConstraint implements ValidatorConstraintInterface {
       return true
     }
 
+    if (typeof value !== 'number') {
```
Comment (robertparkinson) [inline src/main/app/forms/validation/validators/isLowerThan.ts]: refactor as above please

File: src/test/app/forms/validation/validators/amountIsDefinedConstraint.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import { AmountIsDefinedConstraint } from 'app/forms/validation/validators/amountIsDefined'
+
+describe('AmountIsDefinedConstraint', () => {
+  const constraint: AmountIsDefinedConstraint = new AmountIsDefinedConstraint()
+
+  describe('validate', () => {
+    describe('should return true', () => {
+      it('given a value is null', () => {
+        const args: ValidationArguments = {
+          'value': '',
```
Comment (yarpo) [inline src/test/app/forms/validation/validators/amountIsDefinedConstraint.ts]: key of object doesn't have to be wrapped by ' ' as long as it's valid name of variable.

File: src/test/app/forms/validation/validators/minAmountConstraint.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      it('given a value is not a number', () => {
+        const args: ValidationArguments = validationArgs(Amount.MIN_ALLOWED, 'cannot')
+
+        expect(constraint.validate('asdsad', args)).to.equal(false)
+      })
+    })
+  })
+})
+
+function validationArgs (input: number, cannotState: String): ValidationArguments {
```
Comment (yarpo) [inline src/test/app/forms/validation/validators/minAmountConstraint.ts]: `string` not `String`

REQUEST TO CHANGE

File: src/test/app/forms/validation/validators/maxAmountConstraint.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      it('given a value is not a number', () => {
+        const args: ValidationArguments = validationArgs(Amount.MAX_ALLOWED, 'cannot')
+
+        expect(constraint.validate('asdsad', args)).to.equal(false)
+      })
+    })
+  })
+})
+
+function validationArgs (input: number, cannotState: String): ValidationArguments {
```
Comment (yarpo) [inline src/test/app/forms/validation/validators/maxAmountConstraint.ts]: `string` not `String`

TO CHANGE

File: src/main/app/forms/validation/validators/amountIsDefined.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (cannotStateInput === 'cannot' && (value === undefined || value === null || isNaN(value))) {
+      return true
+    }
+
+    return value != null
+  }
+}
+
+export function AmountIsDefined (cannotState: string, validationOptions?: ValidationOptions) {
```
Comment (sabahirfan) [inline src/main/app/forms/validation/validators/amountIsDefined.ts]: As this validator can be used with specific type of amount, i.e. AmountRange. Probably rename it accordingly. Adding some java docs will be helpful as well

File: src/main/app/forms/validation/validators/amountIsNumber.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (cannotStateInput === 'cannot' && (value === null || value === undefined)) {
+      return true
+    }
+
+    return typeof value === 'number'
+  }
+}
+
+export function AmountIsNumber (cannotState: string, validationOptions?: ValidationOptions) {
```
Comment (sabahirfan) [inline src/main/app/forms/validation/validators/amountIsNumber.ts]: Same here please consider this renaming.

File: src/main/app/forms/validation/validators/maxAmount.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    if (cannotStateInput === 'cannot' && (value === null || value === undefined )) {
+      return true
+    }
+
+    return value <= max
+  }
+}
+
+export function MaxAmount (max: number, cannotState: string, validationOptions?: ValidationOptions) {
```
Comment (sabahirfan) [inline src/main/app/forms/validation/validators/maxAmount.ts]: consider renaming please

File: src/main/app/forms/validation/validators/minAmount.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    if (cannotStateInput === 'cannot' && (value === null || value === undefined)) {
+      return true
+    }
+
+    return value >= min
+  }
+}
+
+export function MinAmount (min: number, cannotState: string, validationOptions?: ValidationOptions) {
```
Comment (sabahirfan) [inline src/main/app/forms/validation/validators/minAmount.ts]: consider renaming please

File: src/main/app/forms/models/amount.ts [hunk 10]
```diff
@@ -16,25 +21,25 @@ export class ValidationErrors {
 export class Amount implements Serializable<Amount> {
   static readonly CANNOT_STATE_VALUE = 'cannot'
   static readonly MAX_ALLOWED = 9999999.99
+  static readonly MIN_ALLOWED = 0.01
 
-  @ValidateIf(o => (o.higherValue && o.higherValue > 0 ) || (o.lowerValue && o.lowerValue > 0))
+  @ValidateIf(o => (o.higherValue != null) || (o.lowerValue != null))
```
Comment (yarpo) [inline src/main/app/forms/models/amount.ts]: I'd rather use === and !==

File: src/main/app/forms/validation/validators/isSelectedAlong.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  validate (value: any, args: ValidationArguments) {
+
+    const cannotState = (args.object as any)[args.constraints[0]]
+    const lowerValue = (args.object as any)[args.constraints[1]]
+
+    if (cannotState !== 'cannot') {
+      return true
+    }
+
+    return (lowerValue === null || lowerValue === undefined) && (value === null || value === undefined)
```
Comment (yarpo) [inline src/main/app/forms/validation/validators/isSelectedAlong.ts]: when can this value be `null`?

File: src/main/app/forms/validation/validators/minAmountRange.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+@ValidatorConstraint({ name: 'minAmountRange' })
+export class MinAmountRangeConstraint implements ValidatorConstraintInterface {
+
+  validate (value: any, args: ValidationArguments) {
+
+    const min: number = args.constraints[0]
+    const cannotStateInput = (args.object as any)[args.constraints[1]]
+
+    if (cannotStateInput === 'cannot' && (value === null || value === undefined)) {
```
Comment (yarpo) [inline src/main/app/forms/validation/validators/minAmountRange.ts]: I can see you always check if value is null / undefined. Is there a real scenario that this var can have these values?

**Pull request merged**